mod/state-transition/pkg/core/state_processor_staking.go

```diff
@@ -38,26 +39,36 @@ func (sp *StateProcessor[
 	st BeaconStateT,
 	blk BeaconBlockT,
 ) error {
-	// Verify that outstanding deposits are processed up to the maximum number
-	// of deposits.
-	deposits := blk.GetBody().GetDeposits()
-	index, err := st.GetEth1DepositIndex()
+	// Verify that outstanding deposits matches those listed by contract
+	depositIndex, err := st.GetEth1DepositIndex()
 	if err != nil {
 		return err
 	}
-	eth1Data, err := st.GetEth1Data()
+
+	stateDeposits, err := sp.ds.GetDepositsByIndex(
+		depositIndex,
+		sp.cs.MaxDepositsPerBlock(),
+	)
 	if err != nil {
 		return err
 	}
-	depositCount := min(
-		sp.cs.MaxDepositsPerBlock(),
-		eth1Data.GetDepositCount().Unwrap()-index,
-	)
-	_ = depositCount
-	// TODO: Update eth1data count and check this.
-	// if uint64(len(deposits)) != depositCount {
-	// 	return errors.New("deposit count mismatch")
-	// }
+
+	deposits := blk.GetBody().GetDeposits()
+	if len(stateDeposits) != len(deposits) {
+		return fmt.Errorf("deposits mismatched lengths, state: %d, payload: %d",
+			len(stateDeposits),
+			len(deposits),
+		)
+	}
+
+	for i, sd := range stateDeposits {
+		if !reflect.DeepEqual(sd, deposits[i]) {
```

Comment:
i think we should write a `.Equal(` function on the Deposit struct as well. Not a fan of using DeepEqual especially when the types are interfaces

Reply:
Not as simple as expected to define this given current error types.